src/java.base/share/classes/java/io/FileDescriptor.java

```diff
@@ -372,7 +374,31 @@ private synchronized void beforeCheckpoint() throws CheckpointOpenFileException
         if (valid()) {
             JDKContext ctx = jdk.internal.crac.Core.getJDKContext();
             if (ctx.claimFdWeak(this, this)) {
-                String msg = "FileDescriptor " + this.fd + " left open. ";
+                String path = getPath();
+                String type = getType();
+                String info;
+                if ("socket".equals(type)) {
+                    InetSocketAddress[] addresses = Socket.getAddresses(fd);
+                    StringBuilder sb = new StringBuilder(Socket.getType(fd));
+                    if (addresses == null) {
+                        sb.append("not IPv4/IPv6");
+                    } else {
+                        if (addresses[0] != null) {
+                            sb.append(" local ").append(addresses[0]);
+                        } else {
+                            sb.append(" local not bound");
+                        }
+                        if (addresses[1] != null) {
+                            sb.append(" remote ").append(addresses[1]);
+                        } else {
+                            sb.append(" remote not bound");
+                        }
+                    }
+                    info = sb.toString();
+                } else {
+                    info = (path != null ? path : "unknown path") + " (" + (type != null ? type : "unknown") + ")";
+                }
```

Comment:
I am thinking about this also in a context of `newfd-policies`, where we have to record not only the type & path but also things like current offset etc.
Since this is used only in error handling, I wouldn't mind about the native transitions (or performance in general) too much. If you want, I could create a POJO to cross the native border only once. However this goes a bit against the principle that we want to handle as much as we can in Java rather than native code. That's also the reason why I opted for the formatting in Java even though it might be just as simple to do it in the native.